### DISCUSSION TITLE
Passing delay to device driver

### DISCUSSION THREAD

**Discussion by xiugaze** [stm32-rs/stm32f4xx-hal#622] [General]
Hi,
Is there a proper way implemented to use blocking delay inside of a device driver? There doesn't seem to be a way to create a driver struct that has access to a delay without either 
1. Giving ownership of the delay to the struct
2. Passing a reference to a delay within every API call. 

Clone and copy are not implemented for `stm32f4xx_hal::timer::SYST`. 

I don't know if I am just missing something obvious or the proper way to do this. 

I did look into this [here](https://github.com/rust-embedded/book/issues/246), but I can't find a conclusive answer on how I should be designing drivers.

**Comment by burrbull** [comment]
The second option is much more flexible, so it is more preferable.
You can discuss it in [matrix](https://matrix.to/#/#rust-embedded:matrix.org) if you are not sure.
